### ISSUE TITLE
[Feature Request] Exclude ties in function in_top_k

### ISSUE THREAD

**Issue by gfolego** [tensorflow/tensorflow#10767]
This is a feature request related to #10489 (tie handling in function in_top_k)

Would it be possible to add an option/argument specifying whether ties should be included or excluded?
In other words, make it possible for ties to return False instead of True.

Thanks!

**Comment by aselle** [comment]
@gfolego, could you give some motivation for why this is needed. It seems you worry that there is a trivial null-space to the loss function where you return constant 1's as your logits which causes zero loss since every class is in top_k. However, with random weight initialization this doesn't seem likely to be found with sgd since the gradient will not reliably target that.

**Comment by gfolego** [comment]
@aselle Sure! I'll use ImageNet as an example, but my case is quite similar.
Let's say I have the ILSVRC2017 dataset, and I'm working on image classification only, so my main objective is to get a high top 5 accuracy.
Now, I'm analyzing a CNN's performance given a specific number of training images, which will vary from very few to the complete training set. The motivation behind this analysis is to determine the best trade-off between size and accuracy, so we know how many images we need when collecting a new dataset.

In particular, ILSVRC2017 training dataset (for classification/localization) has between 732 and 1300 images for each class. If I use as few as 5 from each class for training a (large) CNN from scratch, this will most likely result in a CNN with random predictions or maybe predicting always the same class. So the top 5 accuracy for this model will be artificially (and incorrectly, IMHO) high.

So, if I'm early stopping or choosing a model based on this metric, I'll get a CNN that doesn't work in practice. Also, based on this approach, training with 5 samples will most likely achieve better top 5 accuracy than training with the complete set.

Please note that I'm not the first one experiencing this, as indicated in #10489 and in this issue on Keras (using TF's in_top_k) https://github.com/fchollet/keras/issues/6167

TL;DR: If a CNN predicts the same probability for every class, its top 5 accuracy in ILSVRC2017 classification is 100%, which is naturally incorrect.

Thanks!

**Comment by aselle** [comment]
@fchollet, @martinwicke, what is your feeling on this feature? Should we add an argument to in_top_k? It seems like the conceptional conflict is that it is hard to make a top_k that ignores ties in a consistent way but is also fixed size.

**Comment by martinwicke** [comment]
It seems it's hard either way -- if you include ties, you may end up with a larger than wanted set, if you exclude them, you may end up with a smaller than wanted set. If you want to enforce a constant size, you have to pick from the set of ties.

The logic for this would be comparatively complicated, since I assume you'd definitely want to include ties if the set of tied values fits in k (i.e. for top 5, if places 1 and 2 are tied, you still want them in the set).

I think an extra argument `handle_ties` with possible values `include`, `exclude`, or `sample` (the default being sample) would make sense.

**Comment by gfolego** [comment]
@martinwicke Thanks for sharing your thoughts. I agree there's no trivial way to enforce a fixed set size. I'm not familiar with the tensorflow code base, so I'm not exactly sure how functions tf.nn.in_top_k and tf.nn.in_top_k relate to each other, but it seems that they handle ties differently.

Given that, would it be better to change them in order to behave similarly, or would it be better to keep current behaviors? If we were to keep tf.nn.in_top_k's current behavior, I believe we would need the respective parameter option, such as `index` or `order`. Here, I'm assuming your `sample` means random.

Regarding the logic, here's one option (assuming zero index):

```
sort probabilities
if prob[k-1] != prob[k]
    // happy ending
else
    if handle_ties == include
        // return all positions greater than or equal to prob[k-1]
    else if handle_ties == exclude
        // return all positions greater than prob[k-1]
    else if handle_ties == sample or index
        // get all positions greater than prob[k-1] (which will be returned)
        // get all positions equal to prob[k-1] (ties)
        // fill remaining positions (according to sample or index)
```

I hope this helps somehow. But it's a great discussion anyway.
Thanks and best regards!

**Comment by martinwicke** [comment]
I think adding such an option would be a good addition if fixed length top k is asked for, and it would clarify the current behavior.

**Comment by nolanliou** [comment]
@martinwicke I have sent a PR, please take a look if you have time.

**Comment by maxim5** [comment]
Current implementation looks really dangerous. A real case from StackOverflow: [Mysterious ReLu](https://stackoverflow.com/q/48993004/712995)

**Comment by martinwicke** [comment]
I would love a PR to fix this. @nolanliou valiantly tried. A backwards compatible version as described above should not be terribly hard to do.

**Comment by sushreebarsa** [comment]
@gfolego In TensorFlow in_top_k function doesn't have a built-in option to exclude ties. Maybe one workaround can be as follows to Filter Ties Manually by obtaining the indices of the top k elements using in_top_k and Filter out indices where ties occur using additional logic. Please feel free to create a new issue if you still have a concern. Thank you!
